### PR TITLE
IGNITE-23281 Fix for long holding of write lock on Metastorage Compaction

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-23281

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java [hunk 1]
```diff
@@ -1595,4 +1597,9 @@ private boolean isTombstoneForCompaction(byte[] key, long revision) {
             );
         }
     }
+
+    private static WriteOptions createWriteOptions() {
+        // Metastorage recovery is based on the snapshot & external log. WAL is never used for recovery, and can be safely disabled.
+        return new WriteOptions().setDisableWAL(true);
```
Comment (sashapolo) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java:1603]: Can we extract these options into a field?

File: modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/AbstractCompactionKeyValueStorageTest.java [hunk 2]
```diff
@@ -79,6 +92,10 @@ void setUp() {
         assertEquals(List.of(4, 6), collectRevisions(SOME_KEY));
     }
 
+    abstract boolean isPersistent();
```
Comment (sashapolo) [inline modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/AbstractCompactionKeyValueStorageTest.java:95]: Why do you need this method?

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtils.java [hunk 3]
```diff
@@ -64,4 +64,12 @@ public static int indexToCompact(long[] keyRevisions, long compactionRevisionInc
     public static String toUtf8String(byte[] bytes) {
         return new String(bytes, UTF_8);
     }
+
+    /** Asserts that the compaction revision is less than the current repository revision. */
+    public static void assertCompactionRevisionLessCurrent(long compactionRevision, long revision) {
```
Comment (sashapolo) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtils.java:69]:
```suggestion
    public static void assertCompactionRevisionLessThanCurrent(long compactionRevision, long revision) {
```

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtils.java [hunk 4]
```diff
@@ -64,4 +64,12 @@ public static int indexToCompact(long[] keyRevisions, long compactionRevisionInc
     public static String toUtf8String(byte[] bytes) {
         return new String(bytes, UTF_8);
     }
+
+    /** Asserts that the compaction revision is less than the current repository revision. */
```
Comment (sashapolo) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtils.java:68]: What do you mean by "repository revision"?

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java [hunk 5]
```diff
@@ -977,25 +977,27 @@ public void removeWatch(WatchListener listener) {
...
+        try (RocksIterator iterator = index.newIterator()) {
+            iterator.seekToFirst();
 
-        try (WriteBatch batch = new WriteBatch()) {
-            assert revision < rev : String.format(
-                    "Compaction revision should be less than the current: [compaction=%s, current=%s]",
-                    revision, rev
-            );
+            RocksUtils.forEach(iterator, (key, value) -> {
+                rwLock.writeLock().lock();
```
Comment (sashapolo) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java:984]: I'm a little concerned that we take a lock for every key, maybe we should batch it?

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java (hunk 1 above)
Comment (tkalkirill) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java:1603]: I don't know, this object is marked as `java.lang.AutoCloseable`, the documentation doesn't say whether we can close it later or whether it needs to be closed immediately after use. Do you think it's worth putting it in the field?

File: modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/AbstractCompactionKeyValueStorageTest.java (hunk 2 above)
Comment (tkalkirill) [reply modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/AbstractCompactionKeyValueStorageTest.java:95]: We have tests that check something after the storage is restarted. The test implementation is volatile and cannot work with snapshots. I was thinking about how to make the bypass of tests that require working with disks more elegant. Any suggestions?

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtils.java (hunk 3 above)
Comment (tkalkirill) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtils.java:69]: Thanks!

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtils.java (hunk 4 above)
Comment (tkalkirill) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtils.java:68]: Typo, fix it, thanks!

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java (hunk 5 above)
Comment (tkalkirill) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java:984]: Yes, why not.

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java (hunk 1 above)
Comment (sashapolo) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java:1611]: We already do that in similar rocksdb-based storages. There's no point in creating this object every time, just don't forget to close it with the storage

File: modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/AbstractCompactionKeyValueStorageTest.java (hunk 2 above)
Comment (sashapolo) [reply modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/AbstractCompactionKeyValueStorageTest.java:95]: I just didn't see any usages of it, am I missing something?

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java (hunk 1 above)
Comment (tkalkirill) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java:1611]: Ok, why not.

File: modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/AbstractCompactionKeyValueStorageTest.java (hunk 2 above)
Comment (tkalkirill) [reply modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/AbstractCompactionKeyValueStorageTest.java:95]: See `org.apache.ignite.internal.metastorage.server.AbstractCompactionKeyValueStorageTest#testRevisionsAfterRestart`

**Review by sashapolo — APPROVED** [review]

**Pull request merged**